### PR TITLE
beacon/engine: don't omit empty withdrawals in ExecutionPayloadBodies

### PR DESCRIPTION
**Pull request by MariusVanDerWijden**
Fixes an issue where we would omit [] withdrawals.
Adds withdrawals to the ExecutionPayloadBodies tests.

### CODE REVIEW

**Comment by holiman** [comment]
I must be missing something ... but if we don't `omitempty` those, won't empty withdrawals-liss be returned right now, before shanghai, in methods like `GetPayloadBodiesByHashV1`, `GetPayloadBodiesByRangeV1` ?

**Comment by MariusVanDerWijden** [comment]
According to the spec we should return nil pre-shanghai. The hive test suite passes with this change. I'm not sure if we return an empty list or nil though

**Comment by rkrasiuk** [comment]
@holiman according to https://github.com/ethereum/execution-apis/pull/352, `engine_getPayloadBodiesByRangeV1` as well as `ExecutionPayloadBodyV1` type itself are changes introduced with Shanghai. `withdrawals` seems to be a mandatory field in `ExecutionPayloadBodyV1`

File: eth/catalyst/api_test.go [hunk 1]
```diff
@@ -1233,26 +1238,53 @@ func TestNilWithdrawals(t *testing.T) {
 }
 
 func setupBodies(t *testing.T) (*node.Node, *eth.Ethereum, []*types.Block) {
-	genesis, preMergeBlocks := generateMergeChain(10, false)
+	genesis, preMergeBlocks := generateMergeChain(10, true)
 	n, ethservice := startEthService(t, genesis, preMergeBlocks)
+	// enable shanghai on the last block
+	ethservice.BlockChain().Config().ShanghaiTime = &preMergeBlocks[len(preMergeBlocks)-1].Header().Time
```
Comment (holiman) [inline eth/catalyst/api_test.go:1244]: IF shanghai time is `==` that header time, doesn't that mean that _that_ particular header is also shanghai?

File: eth/catalyst/api_test.go [hunk 2]
```diff
@@ -1434,5 +1477,18 @@ func equalBody(a *types.Body, b *engine.ExecutionPayloadBodyV1) bool {
...
+	if a.Withdrawals == nil && b.Withdrawals == nil {
+		return true
+	} else if a.Withdrawals == nil || b.Withdrawals == nil {
+		return false
+	}
+	aBytes, errA = rlp.EncodeToBytes(a.Withdrawals)
+	bBytes, errB = rlp.EncodeToBytes(b.Withdrawals)
+	if errA != errB {
+		return false
+	}
```
Comment (fjl) [inline eth/catalyst/api_test.go:1492]: This check seems kind of long-winded, can't we use `reflect.DeepEqual` here?

**Comment by rjl493456442** [comment]
> @holiman according to [ethereum/execution-apis#352](https://github.com/ethereum/execution-apis/pull/352), `engine_getPayloadBodiesByRangeV1` as well as `ExecutionPayloadBodyV1` type itself are changes introduced with Shanghai. `withdrawals` seems to be a mandatory field in `ExecutionPayloadBodyV1`

It's a bit weird. These two functions are introduced in Shanghai fork, but does it means that these functions can only be called for post-shanghai blocks? If so, should we add some restrictions?

**Comment by MariusVanDerWijden** [comment]
No they can also be called on pre-shanghai blocks

**Pull request merged**